h5io/_h5io.py

```diff
@@ -288,6 +291,7 @@ def _triage_write(
             where + ".csc_matrix_indices",
             cleanup_data=cleanup_data,
             slash=slash,
+            use_state=use_state,
```

Comment:
Is there a case where the `indices` are required to be an object which needs `__getstate__()` to be stored in the HDF5 file?

Reply:
I'll answer here, but it applies to all of these as they're each occuring inside a `... value isinstance(sparse.csX_matrix)` context.

Philosophically, if a user requests `use_state`, who are we to sneakily alter their wishes deeper in the recursive tree?

Empirically, there are three questions:
1. What behaviour do we want to support?
2. How likely is that behaviour?
3. What price are we willing to pay for the support?

For (1), any subclass of the `csX_matrix` will wind up in this branch of the if-else tree. So we can imagine a scenario where somebody has subclassed one of these matrices, and the subclasses uses, e.g., `Foobar` instances instead of `int` to index it (or whatever, I'm not familiar with the class). In that case, when they've explicitly asked for us to `use_state`, I think it's kind to oblige them.

Regard (2), it's extremely likely that this will _never_ happen. I guess these shapes and indices and pointers are pretty standard and not likely to be modified. E.g. subclassing `list` so that it was indexed with bytes instead of integers would be very weird.

But the catch is that the answer to (3) is basically nothing. All we're doing is passing a bool that would otherwise take its default value, so there's no computational expense. If the user is behaving "normally" there is (afaik?) no behaviour difference either way. Finally, if the user is doing something bizarre, it will certainly fail hard without this change, so we can't really get _worse_ than that.

Ultimately, I totally agree it looks a bit funny, but I don't see a reason to not just give the user what they asked for. Never having touched `sparse`, I'm open to the possibility that the flag might actually be harmful in this context, but at first glance that seems unlikely.

Reply:
Yes, I am fine ;-)